PWGCF/Flow/Tasks/flowAnalysisGF.cxx

```diff
@@ -0,0 +1,686 @@
+      }
+    }
+    return;
+  }
+
+  template <datatype dt, typename TCollision, typename TTracks>
+  void processCollision(TCollision collision, TTracks tracks, const float& centrality, const int& field)
+  {
+
+    if (tracks.size() < 1)
+      return;
+    if (centrality < centbinning.front() || centrality > centbinning.back())
+      return;
+    registry.fill(HIST("hEventCount"), 9.5);
+    float vtxz = collision.posZ();
+    fGFW->Clear();
+    fFCpt->ClearVector();
+    float l_Random = fRndm->Rndm();
+
+    for (auto& track : tracks) {
+      ProcessTrack(track, centrality, vtxz, field);
+    }
+
+    FillOutputContainers<dt>((cfgUseNch) ? tracks.size() : centrality, l_Random);
+  }
+
+  template <typename TrackObject>
+  inline void ProcessTrack(TrackObject const& track, const float& centrality, const float& vtxz, const int& field)
+  {
+    float weff = 1, wacc = 1;
+    if constexpr (framework::has_type_v<aod::mctracklabel::McParticleId, typename TrackObject::all_columns>) {
+      if (track.mcParticleId() < 0 || !(track.has_mcParticle()))
+        return;
+
+      auto mcParticle = track.mcParticle();
+      if (!mcParticle.isPhysicalPrimary() || mcParticle.eta() < etalow || mcParticle.eta() > etaup || mcParticle.pt() < ptlow || mcParticle.pt() > ptup || track.tpcNClsFound() < cfgNcls)
+        return;
+
+      if (cfgUseAdditionalTrackCut && !trackSelected(track, field))
+        return;
+
+      if (cfgFillWeights)
+        fWeights->Fill(mcParticle.phi(), mcParticle.eta(), vtxz, mcParticle.pt(), centrality, 0);
+
+      if (!setCurrentParticleWeights(weff, wacc, mcParticle.phi(), mcParticle.eta(), mcParticle.pt(), vtxz))
+        return;
+
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), mcParticle.phi(), mcParticle.eta(), vtxz, wacc);
+      if (cfgFillQA)
+        FillTrackQA<kReco>(track, vtxz);
+
+      FillGFW(mcParticle, weff, wacc);
+    } else if constexpr (framework::has_type_v<aod::mcparticle::McCollisionId, typename TrackObject::all_columns>) {
+      if (!track.isPhysicalPrimary() || track.eta() < etalow || track.eta() > etaup || track.pt() < ptlow || track.pt() > ptup)
+        return;
+
+      if (cfgFillQA)
+        FillTrackQA<kGen>(track, vtxz);
+
+      FillGFW(track, 1., 1.);
+    } else {
+      if (track.tpcNClsFound() < cfgNcls)
+        return;
+
+      if (cfgUseAdditionalTrackCut && !trackSelected(track, field))
+        return;
+
+      if (cfgFillWeights)
+        fWeights->Fill(track.phi(), track.eta(), vtxz, track.pt(), centrality, 0);
+
+      if (!setCurrentParticleWeights(weff, wacc, track.phi(), track.eta(), track.pt(), vtxz))
+        return;
+
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), track.phi(), track.eta(), vtxz, wacc);
+      if (cfgFillQA)
+        FillTrackQA<kReco>(track, vtxz);
+
+      FillGFW(track, weff, wacc);
```

Comment:
I guess you don't need to replicate this code, you already have it in the reco section above
I would suggest the first `if` condition handles the MC generated, its `else` handles the reconstructed with an inner `if` for the handling of the MC reco particularity and the rest in common for data and MC reco

Reply:
Thank you for the suggestion, however, I'm not sure how to implement it correctly because for the data "track" is used and for the MC reconstructed "track.mcParticle()" is used. I couldn't find an easy way to treat both of them in a common statement for all reconstructed without duplicating the code, could it be I'm missing something?